### DISCUSSION TITLE
Problem since replacing Pi 3b with Pi 4; video repeatedly terminates

### DISCUSSION THREAD

**Discussion by robosmith** [GerryDazoo/Slinger#258] [Q&A] [answered]
Here is the log file with no remote inputs starting from a fresh reboot on a server with 2 Tivos configured. Only one (Bob) used.

`Streamer:  Bob Waiting for first stream, flushing any IR requests that arrive while not connected to slingbox
Streamer:  Rick Waiting for first stream, flushing any IR requests that arrive while not connected to slingbox
11/08/2022, 22:59:05   Streaming connection from ('xx.xx.xx.xx', 53062)
Error: Can't find streamer for 12050 in dict_keys(['Bob', 'Rick'])
11/08/2022, 22:59:06   Streaming connection from ('xx.xx.xx.xx', 59101)
Starting remote stream 1

Slinginfo  "Solo/Pro/ProHD" 12 30 192.168.0.26 5001 3000 5
Bob Starting Stream for  xx.xx.xx.xx:59101
Connecting... ('192.168.0.26', 5001) Control
Bob Selecting Video Source 3
Keycodes= [4, 5, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 24, 26, 27, 28, 29, 32, 35, 38, 39, 40, 41, 42, 46, 50, 62, 63, 64, 65, 66, 67, 68, 69, 70, 74, 75, 76, 77]
VideoParameters: Resolution= 12 FrameRate= 30 VideoBandwidth= 2000 VideoSmoothness= 63 IframeRate= 5 AudioBitRate= 64
Connecting... ('192.168.0.26', 5001) Stream
Bob Stream started at 11/08/2022, 22:59:13  733 0
11/08/2022, 22:59:20   Streaming connection from ('xx.xx.xx.xx', 54934)
Bob Got Streamer Control Message STREAM
Starting remote stream 2
Bob New Stream Starting 0
........11/08/2022, 23:00:41  Bob Stream Terminated for  xx.xx.xx.xx:54934 [Errno 104] Connection reset by peer
1 active remote connections
11/08/2022, 23:00:41   Streaming connection from ('xx.xx.xx.xx', 39413)
Bob Got Streamer Control Message STREAM
Starting remote stream 2
Bob New Stream Starting 0
11/08/2022, 23:00:43  Bob 2 Clients.xx.xx.xx.xx:59101 xx.xx.xx.xx:39413 
........11/08/2022, 23:02:03  Bob Stream Terminated for  xx.xx.xx.xx:39413 [Errno 104] Connection reset by peer
1 active remote connections
.11/08/2022, 23:02:03   Streaming connection from ('xx.xx.xx.xx', 58720)
Bob Got Streamer Control Message STREAM
Starting remote stream 2
Bob New Stream Starting 0
11/08/2022, 23:02:13  Bob 2 Clients.xx.xx.xx.xx:59101 xx.xx.xx.xx:58720 
........11/08/2022, 23:03:26  Bob Stream Terminated for  xx.xx.xx.xx:58720 [Errno 104] Connection reset by peer
1 active remote connections
11/08/2022, 23:03:27   Streaming connection from ('xx.xx.xx.xx', 50457)
Bob Got Streamer Control Message STREAM
Starting remote stream 2
Bob New Stream Starting 0
.11/08/2022, 23:03:43  Bob 2 Clients.xx.xx.xx.xx:59101 xx.xx.xx.xx:50457 
.......11/08/2022, 23:04:49  Bob Stream Terminated for  xx.xx.xx.xx:50457 [Errno 104] Connection reset by peer
1 active remote connections
11/08/2022, 23:04:49   Streaming connection from ('xx.xx.xx.xx', 35404)
Bob Got Streamer Control Message STREAM
Starting remote stream 2
Bob New Stream Starting 0
.11/08/2022, 23:04:54  Bob Stream Terminated for  xx.xx.xx.xx:35404 [Errno 11] Resource temporarily unavailable
1 active remote connections
11/08/2022, 23:04:57   Streaming connection from ('xx.xx.xx.xx', 46392)
Bob Got Streamer Control Message STREAM
Starting remote stream 2
Bob New Stream Starting 0
.11/08/2022, 23:05:13  Bob 2 Clients.xx.xx.xx.xx:59101 xx.xx.xx.xx:46392 
.......11/08/2022, 23:06:19  Bob Stream Terminated for  xx.xx.xx.xx:46392 [Errno 104] Connection reset by peer
1 active remote connections
11/08/2022, 23:06:20   Streaming connection from ('xx.xx.xx.xx', 39394)
Bob Got Streamer Control Message STREAM
Starting remote stream 2
Bob New Stream Starting 0
..11/08/2022, 23:06:43  Bob 2 Clients.xx.xx.xx.xx59101 xx.xx.xx.xx:39394 
......11/08/2022, 23:07:42  Bob Stream Terminated for  xx.xx.xx.xx39394 [Errno 32] Broken pipe
1 active remote connections
11/08/2022, 23:07:42   Streaming connection from ('xx.xx.xx.xx, 45129)
Bob Got Streamer Control Message STREAM
Starting remote stream 2
Bob New Stream Starting 0
...11/08/2022, 23:08:13  Bob 2 Clients.xx.xx.xx.xx:59101 xx.xx.xx.xx:45129 
.....11/08/2022, 23:09:03  Bob Stream Terminated for  xx.xx.xx.xx:45129 [Errno 104] Connection reset by peer
1 active remote connections
11/08/2022, 23:09:03   Streaming connection from ('xx.xx.xx.xx', 59538)
Bob Got Streamer Control Message STREAM
Starting remote stream 2
Bob New Stream Starting 0
....11/08/2022, 23:09:43  Bob 2 Clients.xx.xx.xx.xx:59101 xx.xx.xx.xx:59538 
...11/08/2022, 23:10:06  Bob Stream Terminated for  xx.xx.xx.xx:59101 [Errno 11] Resource temporarily unavailable
1 active remote connections
11/08/2022, 23:10:13  Bob Stream Terminated for  xx.xx.xx.xx:59538 [Errno 104] Connection reset by peer
0 active remote connections
Bob Shutting down connections
Streamer:  Bob Waiting for first stream, flushing any IR requests that arrive while not connected to slingbox

I don't understand why the stream keeps getting terminated and the cycle eventually shuts down and stops reconnecting.
`

**Comment by GerryDazoo** [comment, marked as answer]
Firstly it's your media player that's shutting down the connection right away. In your case almost right away.

11/08/2022, 22:59:20 Streaming connection from ('76.167.65.110', 54934)
11/08/2022, 23:00:41 Bob Stream Terminated for 76.167.65.110:54934 [Errno 104] Connection reset by peer

Check  the logs if there are any for your mediaplayer to see why it's complaining. VLC? Running on a PI4 instead of a 3 should not be an issue.

You didn't post ALL the server output. There is important info in the startup.

**Reply by robosmith** [reply]
Using MxPlayer on Fire TV stick. Thanks for the explanation. I don't know where to look for MxPlayer log.
It seems to be working reliably this morning, though it did shutdown a couple of times earlier, so now I'm even more confused.
And just now after playing continuously for a few hours, it terminated the video again.

```
Bob Got Streamer Control Message IR
Bob Sending IR keycode 66 3 for xx.xx.xx.xx
.......11/09/2022, 15:08:27  Bob 1 Clients.xx.xx.xx.xx:33605 
.........11/09/2022, 15:09:57  Bob 1 Clients.xx.xx.xx.xx:33605 
.........11/09/2022, 15:11:27  Bob 1 Clients.xx.xx.xx.xx:33605 
.........11/09/2022, 15:12:57  Bob 1 Clients.xx.xx.xx.xx:33605 
.........11/09/2022, 15:14:27  Bob 1 Clients.xx.xx.xx.xx:33605 
.........11/09/2022, 15:15:57  Bob 1 Clients.xx.xx.xx.xx:33605 
.....11/09/2022, 15:16:47  Bob Stream Terminated for  xx.xx.xx.xx:33605 [Errno 11] Resource temporarily unavailable
0 active remote connections
Bob Shutting down connections
Streamer:  Bob Waiting for first stream, flushing any IR requests that arrive while not connected to slingbox

```

**Reply by robosmith** [reply]
Here is the initial startup log. It includes my modification of the code to keep trying to connect with the slingbox 10x at the address in the config.ini because of problems with the code for finding the IP after a failure. See "iter =" log entries

```
Version : 3.08 Running on Linux-5.15.74-v8+-aarch64-with-glibc2.31 pid= 621
Using config file config.ini
Connection Manager Running on port 12050 with 10 max streams using URL slingbox.
BOXES [('sb1', 'Bob'), ('sb2', 'Rick')]
Streamer Running:  10 config.ini Bob Bob 12050 8388608
starting up on  port 12050 Checking for slingbox at8388608 
('192.168.0.26', 5001)
iterStreamer Running:   010
 Error connecting to config.ini  ('192.168.0.26', 5001)
Rick Rick 12050 8388608
Checking for slingbox at ('192.168.0.6', 5002)
iter 0
Error connecting to  ('192.168.0.6', 5002)
Checking for slingbox at ('192.168.0.26', 5001)
iter 1
Error connecting to  ('192.168.0.26', 5001)
Checking for slingbox at ('192.168.0.6', 5002)
iter 1
Error connecting to  ('192.168.0.6', 5002)
Checking for slingbox at ('192.168.0.26', 5001)
iter 2
Error connecting to  ('192.168.0.26', 5001)
Checking for slingbox at ('192.168.0.6', 5002)
iter 2
Error connecting to  ('192.168.0.6', 5002)
Building page for Bob
Reading Custom Remote definition from Tivo.Premiere.Remote.txt
Building page for Rick
Reading Custom Remote definition from Tivo.Premiere.Remote.txt
 * Serving Flask app "slingbox_server" (lazy loading)
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: off
Checking for slingbox at ('192.168.0.26', 5001)
Checking for slingbox at ('192.168.0.6', 5002)
('192.168.0.26', 5001) OK
Bob('192.168.0.6', 5002)  OKUsing slingbox at 
 ('192.168.0.26', 5001)
Rick Using slingbox at  ('192.168.0.6', 5002)
Streamer:  Bob Waiting for first stream, flushing any IR requests that arrive while not connected to slingbox
Streamer:  Rick Waiting for first stream, flushing any IR requests that arrive while not connected to slingbox

```

**Reply by GerryDazoo** [reply]
ROBO. Two things
1) You should get the latest code. 
2) That reason code for the stream dropping "resource unavailable" means the server ran out of TCP buffer space. It implies either network congestion or the client side isn't keeping up with the video stream. 
So, just checking, are you sure you bumped up the TCP buffers per the readme on the new Pi4.

**Reply by robosmith** [reply]
Ok, got the latest code. Thanks for that. 
Also, my friend who set it up was having permission problem with  bumping up the TCP buffers, but that is now done.
So far, so good.

**Reply by DoogieHowitzerWMD** [reply]
> ROBO. Two things
> 
> 1. You should get the latest code.
> 2. That reason code for the stream dropping "resource unavailable" means the server ran out of TCP buffer space. It implies either network congestion or the client side isn't keeping up with the video stream.
>    So, just checking, are you sure you bumped up the TCP buffers per the readme on the new Pi4.

Is this what is bumping up the TCP buffer?

`Add or modify the two following lines to your /etc/sysctl.conf `
`    net.core.rmem_max = 8192000`
`    net.core.wmem_max = 8192000`

Using `sudo nano /etc/sysctl.conf` I haven't been able to get that file to accept those additional lines. Should I try typing this in terminal instead?

`sysctl -w net.core.rmem_max=8192000`
`sysctl -w net.core.wmem_max=8192000`

**Reply by GerryDazoo** [reply]
you need to be root to do that so sudo ....
I'm not sure those commands make the changes permanent. Give it a go and then check your sysctl.conf file to see if the entries are there.

**Reply by DoogieHowitzerWMD** [reply]
> you need to be root to do that so sudo .... I'm not sure those commands make the changes permanent. Give it a go and then check your sysctl.conf file to see if the entries are there.

I tried writing to the file again and was able to do so this time. I had been using sudo previously, but I don't remember the exact verbage it gave me for why the file could not be saved. It could have been some sort of typo or me forgetting to turn off the write protections (not sure if read-only boot partition would affect system files of this sort or if it exclusively applies to what is in /boot). In any case, I was able to add it this time, and after rebooting, I ran `sysctl -a` and scrolling up to verify the settings were there.

That may have been the source of my issues that you had been helping me to troubleshoot yesterday in another discussion thread. Since increasing the buffer size, I see that I have streams up from 3 slingbox 500's, and my first gen Pi running the server is clocking at 60-70% cpu usage from that. I feel like I could still potentially give pypy a try to see if I can reduce that load, as I'd feel bad running that old Pi into the ground.

**Reply by GerryDazoo** [reply]
It would be nice to see how pypy behaves but don't worry about running the py into the ground. It may as well do something useful instead of just running the OS idle loop.

**Reply by DoogieHowitzerWMD** [reply]
My concern is that I will add a fourth slingbox to the equation (I'm short a power adapter atm), so I'll be pushing that Pi a little harder. I have the cpu heatsinked, but the thought for providing adequate ventilation with the Pi case designs just wasn't there at that time - the original Pi was entirely geared and marketed towards teaching children about computers.

Regardless, I know the server is working fine, because the problems I am seeing now are from how VLC is handling the streams. I am seeing VLC incrementally increase the size of the cache every 15 minutes or so, which means it must be running up against the buffer. From what I gather, I need to set higher values for the network cache (for receiving the packets from the stream) and the cache for the output multiplexer (for the time between whatever processes VLC is performing on the received packets and producing the output). I'm confident I can do that through VLC's gui, but I don't have VNC configured for any of my dedicated Pi streamers - only SSH.

**Comment by DoogieHowitzerWMD** [comment]
Are you running the server on the Pi 4 or the VLC stream?

**Reply by robosmith** [reply]
The Dazoo server code is running on the Pi 4. Client MxPlayer is running on Fire TV stick.
